### ISSUE TITLE
Load balancing in JetStream Work Queue

### ISSUE THREAD

**Issue by ds3p2** [nats-io/nats.js#493]
I am trying to write a demo for the JetStream Work Queue with push consumer. I have some question about the load balance behavior I observed.  

My environment:
- client: nats@2.6.1
- server: nats-server 2.7.4 (using docker)

My producer code:
```typescript
import { connect, JetStreamClient, JSONCodec, RetentionPolicy } from "nats";

async function main() {
  const nc = await connect({ servers: "127.0.0.1" });

  const jsm = await nc.jetstreamManager();

  await jsm.streams.add({
      name:"test",
      subjects: ["test"],
      retention: RetentionPolicy.Workqueue,
  });

  setInterval(dispatch.bind(nc.jetstream()), 1000);
}

async function dispatch(this: JetStreamClient) {
  const encoder = JSONCodec();
  const ack = await this.publish(
    SUBJECT,
    encoder.encode({ task: `sample task at ${new Date().toISOString()}` })
  );
  console.log(`message sent with seq number ${ack.seq}`);
}

main().then(
  () => {},
  (err) => console.error(err)
);

```

My consumer code: 
```typescript
import {
  connect,
  consumerOpts,
  createInbox,
  JsMsg,
  JSONCodec,
  NatsError,
} from "nats";
import util from "util";

async function main(pid: string) {
  console.log("start worker: ", pid);
  const nc = await connect({ servers: "127.0.0.1" });

  const js = nc.jetstream();

  const opt = consumerOpts();
  opt.manualAck();
  opt.ackExplicit();
  opt.queue("x");
  opt.callback(handler);
  opt.maxAckPending(2);
  opt.deliverAll();
  opt.deliverTo(createInbox("x"));
  opt.durable("x");

  await js.subscribe("test", opt);
}

async function handler(err: NatsError, msg: JsMsg) {
    const jc = JSONCodec();
    console.log("message get:", msg.seq);
    await doWork(jc.decode(msg.data));
    msg.ack();
}

async function doWork(data: unknown): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, 2000));
}

main(process.pid.toString()).then(
  () => {},
  (err) => console.error(err)
);

```

If I run the consumer code twice in the terminal, I have two nodejs process, each of them subscribe the work queue through the same consumer `x`.  
What I expected to see is that the content created in the producer being dispatched to the two processes evenly.  
But what I see in the log is a random dispatch: 

Log for consumer process 1:
```
start worker:  1323710
message get: 43
message get: 44
message get: 45
message get: 46
message get: 48
message get: 49
message get: 50
message get: 51
message get: 55
message get: 58
message get: 60
message get: 62
message get: 65
message get: 67
message get: 69
message get: 72
message get: 73
message get: 75
message get: 76
message get: 78
message get: 81
message get: 87
message get: 88
message get: 89
message get: 91
message get: 92
message get: 94
```

Log for consumer process 2: 
```
start worker:  1323822
message get: 47
message get: 52
message get: 53
message get: 54
message get: 56
message get: 57
message get: 59
message get: 61
message get: 63
message get: 64
message get: 66
message get: 68
message get: 70
message get: 71
message get: 74
message get: 77
message get: 79
message get: 80
message get: 82
message get: 83
message get: 84
message get: 85
message get: 86
message get: 90
message get: 93
message get: 95
message get: 96
```

I read from the [document](https://docs.nats.io/nats-concepts/core-nats/queue#queue-groups) that in NATS core the queue group does a random selection. But I did not find respective description about JetStream.  
Is JetStream also doing a random selection? Is there any config so that it can have more predictive behavior such as:
- round-robin
- subscriber with more pending ack message has lower priority for dispatching

**Comment by aricart** [comment]
The push consumer load balancing is dictated by the server - if you don't set up some sort of limit on messages, the nats server will send until it decides to send to someone else. With the very small ack pending you set, what you get is the server sending 2 messages to a client and nothing to the other. And depending on your luck, it will divide the work less than optimal.

Instead, you may want to consider a pull consumer with the same 2 messages for the request; then, the load will be distributed as your worker becomes available - so remove the ackPending limit, and on the pull request, try to get 5 messages; you will likely get more division of work. Also, note that because you are waiting on the callback, if your `doWork()` won't scale properly - try simply dispatching to the do work, and have that function ack the message. Your time to complete that test will probably be about 2 seconds for the entire batch.

**Comment by aricart** [comment]
Please reopen if you have more questions

**Comment by ds3p2** [comment]
Hi @aricart 

I think I should make my use case more clear. I am not working on a high throughput case. 
What I am trying to demo is something like a task queue, the producer send tasks to the queue, and multiple workers subscribe to the queue to process the task.  
The reason I set the `maxAckPending` to be equal to the number of workers is that worker is supposed to do only one task at a time.
What I expected is something like: 
- When message 1 arrives, it will be randomly dispatched to worker 1, and this message will keep not being ack-ed
- When message 2 arrives, NATS will found that worker 1 has un-ack-ed message while worker 2 does not. So the message will be dispatched to worker 2. Now message 2 will also be pending
- When message 3 arrives, it will not be dispatched because `maxAckPending` exceeds. Once any of the worker finishes it processing and ack the message, message 3 will be dispatched to that freed worker.

Regarding your comments:
- As I mentioned, my `doWork` is not supposed to scale at all - In my case one worker should only handle one task at a time, and the message will not get ack-ed until the whole processing is completed, the processing time could be arbitrary long (like hours).  
- I read that the pull consumer needs to acquire message on demand instead of being notified by the server. If I use the pull consumer, I will need to poll the NATS server when there is no message in the queue. I am not sure if it is possible to do long-poll. If not, it means there will be delay between message arrival and consumption (according to the poll interval). This is what I am trying to avoid. 
- I read the document of `flowControl`, I am not sure what it controls using the sliding-window protocol. If it is about the message number being sent in batch, then as I mentioned the number should always be 1.

Please let me know if you have more guide on how to fulfill my requirement.

**Comment by ds3p2** [comment]
@aricart  It seems I am not privileged to reopen the issue

**Comment by aricart** [comment]
so one thing with the max-ack pending is that they are shared across all queue members. So in that case you will have to go with a pull subscriber - you may want to then use the API here. it won't be efficient, but at least it will be responsive to your worker - you'll only pull when you can work on one, and you can scale horizontally as much as you want. 

https://github.com/nats-io/nats.deno/blob/main/jetstream.md#requesting-a-single-message

**Comment by aricart** [comment]
If you want a long poll, you can use:

https://github.com/nats-io/nats.deno/blob/main/jetstream.md#fetching-batch-of-messages

with a batch of one, the iterator remains open for the duration of the poll so you'll get a message or not, but when the iterator finishes, you can poll again.

**Comment by ds3p2** [comment]
Hi @aricart 

I tried with the `fetch` API and it works well. I think long poll will be a suitable solution.  
Just one more question, is there any restriction on the `expires` option in `fetch` API? If I set this value long enough, I think it will somehow looks like a “push mode”.

**Comment by aricart** [comment]
you will only get up to the batch size, the poll just gives more opportunity for you to get a message if there's nothing at the time of the request.

**Comment by ds3p2** [comment]
I mean the following two approaches look similar at least from the consumer side:
- push subscribe with `maxAckPending(1)`
- long poll with `fetch(STREAM, DURABLE, { batch: 1, expires: Infinity })`

What I want to know is if there is limitation on the `expires` options, can I set it to any large (such as `Infinity` or `Number.MAX_VALUE`)?

**Comment by derekcollison** [comment]
They look similar on the surface, but the poll allows N number of apps to all be doing the same thing by default. The push model can be made to be similar by setting max ack pending to a number of parallel jobs you want to be doing at a time and binding the consumer to a queue group.

For pull, the consumer can be configured to limit aspects of a pull request.

```go
	// Pull based options.
	MaxRequestBatch   int           `json:"max_batch,omitempty"`
	MaxRequestExpires time.Duration `json:"max_expires,omitempty"`
```

**Comment by derekcollison** [comment]
That snippet is Go, but hopefully makes sense. It's part of the consumer configuration.

**Comment by ds3p2** [comment]
Hi @derekcollison 

Thanks for the info, I think the respective API in JS client should be `maxPullBatch` `maxPullRequestExpires`. Please correct me if I am wrong.

Besides, your description about push mode is exactly what I tried to do in the origin post. However I was told that push mode does not work like this so I switched to long poll. Do you mean that I can also implement it with push mode?

**Comment by derekcollison** [comment]
Push mode can work, just more care and feeding and we have seen folks get bit by flow control issues blowing up their apps.

So if you know you want max N requests being worked on in parallel, you can set MaxAckPeding to that, with AckExplicit and bind the consumer to a queue group.

Right now IIRC you can not update a consumer, so with this setup you can't easily scale up where as pull is dynamic, so the apps could be put into a simple auto-scaler etc.

**Comment by ds3p2** [comment]
Hi @derekcollison 

> So if you know you want max N requests being worked on in parallel, you can set MaxAckPeding to that, with AckExplicit and bind the consumer to a queue group.

I have tried what you said in my original code snippet. I defined a consumer with MaxAckPeding=2 and AckExplicit=true and bound it to a work queue. What I want is to launch two worker process using the same consumer and messages can be load balanced to these two workers.  
What I observed is that messages cannot be dispatched evenly on the processes. Sometimes one process will have 2 un-ack-ed messages while the other process is idle. 

> Right now IIRC you can not update a consumer

I noticed an API `jsm.consumers.update` but not yet tried with that. Isn't it for updating a consumer?

**Comment by derekcollison** [comment]
We randomize on selection for DQ, so 2 will not give you a decent sample with a small number of messages. If you do 2, but run 10_000 msgs through it should be pretty well distributed.

I am not sure you can update MaxAckPending for the consumer yet. Folks have asked for it, just a matter of priorities atm.

**Comment by ds3p2** [comment]
Hi @derekcollison 

> We randomize on selection for DQ

Yes randomization is what I was asking about at the very beginning. Is there any configure about this selection behavior? I would like something more "stable":
- round-robin
- subscriber with more pending ack message has lower priority for dispatching

> I am not sure you can update MaxAckPending for the consumer yet. Folks have asked for it, just a matter of priorities atm.

I just had a try, it seems to work, no error thrown and I can confirm the modification using `nats` cli.

**Comment by derekcollison** [comment]
No, at core NATS level we have no semantics to be smarter, and while I was working at Google I saw the power of not trying to be too smart and just using random selection at scale.

With pull consumers you are essentially in a bit more control and present a FIFO pattern here, since each app has to send a request to get a message or batch of messages.

**Comment by ds3p2** [comment]
I see. I think pull consumer with long-poll is sufficient for my need so far.  
Thanks for the support from all of you.  
I am closing this issue.
